server/src/main/java/org/elasticsearch/index/codec/vectors/ES816BinaryQuantizedVectorsWriter.java

```diff
@@ -0,0 +1,987 @@
+    private static final long SHALLOW_RAM_BYTES_USED = shallowSizeOfInstance(ES816BinaryQuantizedVectorsWriter.class);
+
+    private final SegmentWriteState segmentWriteState;
+    private final List<FieldWriter> fields = new ArrayList<>();
+    private final IndexOutput meta, binarizedVectorData;
+    private final FlatVectorsWriter rawVectorDelegate;
+    private final ES816BinaryFlatVectorsScorer vectorsScorer;
+    private boolean finished;
+
+    /**
+     * Sole constructor
+     *
+     * @param vectorsScorer the scorer to use for scoring vectors
+     */
+    protected ES816BinaryQuantizedVectorsWriter(
+        ES816BinaryFlatVectorsScorer vectorsScorer,
+        FlatVectorsWriter rawVectorDelegate,
+        SegmentWriteState state
+    ) throws IOException {
+        super(vectorsScorer);
+        this.vectorsScorer = vectorsScorer;
+        this.segmentWriteState = state;
+        String metaFileName = IndexFileNames.segmentFileName(
+            state.segmentInfo.name,
+            state.segmentSuffix,
+            ES816BinaryQuantizedVectorsFormat.META_EXTENSION
+        );
+
+        String binarizedVectorDataFileName = IndexFileNames.segmentFileName(
+            state.segmentInfo.name,
+            state.segmentSuffix,
+            ES816BinaryQuantizedVectorsFormat.VECTOR_DATA_EXTENSION
+        );
+        this.rawVectorDelegate = rawVectorDelegate;
+        boolean success = false;
+        try {
+            meta = state.directory.createOutput(metaFileName, state.context);
+            binarizedVectorData = state.directory.createOutput(binarizedVectorDataFileName, state.context);
+
+            CodecUtil.writeIndexHeader(
+                meta,
+                ES816BinaryQuantizedVectorsFormat.META_CODEC_NAME,
+                ES816BinaryQuantizedVectorsFormat.VERSION_CURRENT,
+                state.segmentInfo.getId(),
+                state.segmentSuffix
+            );
+            CodecUtil.writeIndexHeader(
+                binarizedVectorData,
+                ES816BinaryQuantizedVectorsFormat.VECTOR_DATA_CODEC_NAME,
+                ES816BinaryQuantizedVectorsFormat.VERSION_CURRENT,
+                state.segmentInfo.getId(),
+                state.segmentSuffix
+            );
+            success = true;
+        } finally {
+            if (success == false) {
+                IOUtils.closeWhileHandlingException(this);
+            }
+        }
+    }
+
+    public FlatFieldVectorsWriter<?> addField(FieldInfo fieldInfo) throws IOException {
+        FlatFieldVectorsWriter<?> rawVectorDelegate = this.rawVectorDelegate.addField(fieldInfo);
+        if (fieldInfo.getVectorEncoding().equals(VectorEncoding.FLOAT32)) {
+            @SuppressWarnings("unchecked")
+            FieldWriter fieldWriter = new FieldWriter(fieldInfo, (FlatFieldVectorsWriter<float[]>) rawVectorDelegate);
+            fields.add(fieldWriter);
+            return fieldWriter;
+        }
+        return rawVectorDelegate;
```

Comment:
For `VectorEncoding` different form `FLOAT32` what is the benefit of delegating to rawVectorDelegate instead of throwing an unsupported exception?

Reply:
Simplicity in testing & this is what Lucene will do. I don't want to deviate too much from what the format will be in Lucene.